File: pronacej-backend-seguridad/src/main/java/net/latinus/sistema/integral/gestion/seguridad/service/IA/InformeEgresoPIIService.java
package net.latinus.sistema.integral.gestion.seguridad.service.IA;

import jakarta.servlet.http.HttpServletRequest;
import net.latinus.sistema.integral.gestion.seguridad.model.both.BodyEncriptado;
import net.latinus.sistema.integral.gestion.seguridad.model.both.InformeEgresoPIIDTO;
import net.latinus.sistema.integral.gestion.seguridad.model.response.PaginacionResponse;
import net.latinus.sistema.integral.gestion.seguridad.model.response.RespuestaPorDefectoAuditoria;

public interface InformeEgresoPIIService {
    
    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data InformeEgresoPIIDTO Paginada
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado PaginacionRequest datos para obtener todos los informes de egreso.
     *
     * @return RespuestaPorDefectoAuditoria<PaginacionResponse<InformeEgresoPIIDTO>>
     */
    RespuestaPorDefectoAuditoria<PaginacionResponse<InformeEgresoPIIDTO>> obtenerInformesEgreso(
        HttpServletRequest httpServletRequest, BodyEncriptado bodyEncriptado);
    
    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data InformeEgresoPIIDTO si el informe se creó con éxito
     *
     * @param httpServletRequest request petición.
     * @param bodyEncriptado objeto InformeEgresoPIIDTO a crear.
     *
     * @return RespuestaPorDefectoAuditoria<InformeEgresoPIIDTO>
     */
    RespuestaPorDefectoAuditoria<InformeEgresoPIIDTO> crearInformeEgreso(
        HttpServletRequest httpServletRequest, BodyEncriptado bodyEncriptado);
    
    /**
     * Elimina un informe de egreso del sistema
     *
     * @param httpServletRequest HttpServletRequest datos del request.
     * @param bodyEncriptado InformeEgresoPIIDTO datos del informe a eliminar.
     *
     * @return RespuestaPorDefectoAuditoria<Boolean>
     */
    RespuestaPorDefectoAuditoria<Boolean> eliminarInformeEgreso(
        HttpServletRequest httpServletRequest, BodyEncriptado bodyEncriptado);
}

File: pronacej-front/src/app/modules/seguridad/services/informeEgresoPII.service.ts
import { Injectable } from '@angular/core';
import { InformeEgresoPIIDTO } from 'app/core/model/both/informeEgresoPIIDTO.model';
import { PaginacionRequest } from 'app/core/model/request/PaginacionRequest.model';
import { PaginacionResponse } from 'app/core/model/response/PaginacionResponse.model';
import { RespuestaPorDefecto } from 'app/core/model/response/RespuestaPorDefecto.model';
import { BackendService } from 'app/core/services/backend.service';
import { Observable } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class InformeEgresoPIIService {
    private path = '/informe-egreso-pii';

    constructor(private backendService: BackendService) {}

    /**
     * Obtiene los informes de egreso PII disponibles para el sistema de manera paginada
     *
     * @param paginacionRequest PaginacionRequest objeto con los parámetros de paginación
     * @param nemonicoMenu string nemonico del menú del sistema
     *
     * @returns Observable<RespuestaPorDefecto<PaginacionResponse<InformeEgresoPIIDTO>>>
     */
    obtenerInformesEgresoPaginado(
        paginacionRequest: PaginacionRequest,
        nemonicoMenu = ''
    ): Observable<
        RespuestaPorDefecto<PaginacionResponse<InformeEgresoPIIDTO>>
    > {
        let endPoint = this.path + '/obtenerInformesEgresoPaginado';
        return this.backendService.postFinal(
            endPoint,
            paginacionRequest,
            nemonicoMenu
        );
    }

    /**
     * Crea un informe de egreso PII en el sistema con los datos enviados en el request
     *
     * @param informeEgresoDTO InformeEgresoPIIDTO datos del informe de egreso a crear
     * @param nemonicoMenu string nemonico del menú del sistema
     *
     * @returns Observable<RespuestaPorDefecto<InformeEgresoPIIDTO>>
     */
    crearInformeEgreso(
        informeEgresoDTO: InformeEgresoPIIDTO,
        nemonicoMenu: string = ''
    ): Observable<RespuestaPorDefecto<InformeEgresoPIIDTO>> {
        let endPoint = this.path + '/crearInformeEgreso';
        return this.backendService.postFinal(
            endPoint,
            informeEgresoDTO,
            nemonicoMenu
        );
    }

    /**
     * Elimina un informe de egreso PII en el sistema
     *
     * @param informeEgresoDTO InformeEgresoPIIDTO datos del informe de egreso a eliminar
     * @param nemonicoMenu string nemonico del menú del sistema
     *
     * @returns Observable<RespuestaPorDefecto<boolean>>
     */
    eliminarInformeEgreso(
        informeEgresoDTO: InformeEgresoPIIDTO,
        nemonicoMenu = ''
    ): Observable<RespuestaPorDefecto<boolean>> {
        let endPoint = this.path + '/eliminarInformeEgreso';
        return this.backendService.postFinal(
            endPoint,
            informeEgresoDTO,
            nemonicoMenu
        );
    }

    /**
     * Verifica y maneja los errores generales del servicio
     *
     * @param error any error a verificar
     * @param mostrarError boolean indica si se debe mostrar el error
     * @returns any
     */
    async checkError(error: any, mostrarError = true): Promise<string> {
        return await this.backendService.checkError(error, mostrarError);
    }
}
